### PR TITLE
Captcha

### PR DESCRIPTION
**Pull request by nogaems**
add: /captcha endpoint

There are two ways to use this endpoint:
1) Use GET request to obtain a captcha, use POST request to send its code and get a token, put that token in request headers as 'X-Captcha-Token: <value>' and you will be allowed to perform a request on endpoints requiring captcha authorization. After single usage this captcha becomes unusable. The token has relatively short expiration time, but this time is enough to send a form.
2) Send captcha structure along with the rest request body as a part of json object with the following syntax:
{...{'captcha': {'id': <value>, 'code': <value>}}...}
Note that the first method is prioritized over the second one.
Also, you can use PUT request for testing purposes. It does nothing besides consuming provided captcha and returning status code 200 if everything went correct, otherwise it tells what's happened.

### CODE REVIEW

File: src/auth.py [hunk 1]
```diff
@@ -34,14 +38,31 @@ def hashed_password(password: str, salt: str = None) -> (str, str):
     It is possible to specify the life time of a token in seconds with exp_time.
     Default value is an hour.
     '''
-    payload = {
-        'sub': username,
-        'exp': datetime.datetime.utcnow() + datetime.timedelta(seconds=exp_time),
-    }
+    payload = {'exp': datetime.datetime.utcnow(
+    ) + datetime.timedelta(seconds=exp_time)}
```
Comment (gordon-quad) [inline src/auth.py:42]: I guess it would be nicer to have just a timestamp here, not whole date string

Comment (gordon-quad) [reply src/auth.py:42]: just a thought

File: src/auth.py [hunk 2]
```diff
@@ -54,18 +75,20 @@ def hashed_password(password: str, salt: str = None) -> (str, str):
...
 
-    if 'sub' not in decoded:
-        return (False, 'Token subject is missing')
-
-    user = session.query(model.User).filter_by(username=decoded['sub']).first()
-    if not user:
-        return (False, 'Token subject doesn\'t exist')
+    missing = [field for field in required_fields if field not in decoded]
+    if missing:
+        return (False, f'Token fields are missing: {repr(missing)}')
```
Comment (gordon-quad) [inline src/auth.py:88]: I'm not sure if it's a good idea to report if something is missing from encrypted payload that your app itself generated, possible data leak.

File: src/util.py [hunk 3]
```diff
@@ -23,3 +25,55 @@ def strip_list(data: dict) -> dict:
     for k, v in data.items():
         data[k] = v if isinstance(v, str) else v[0]
     return data
+
+
+def check_id(request, decoded):
+    if decoded['id'] in request.app.captcha.storage:
```
Comment (gordon-quad) [inline src/util.py]: please make separate API in faptcha for this operation

**Review by gordon-quad — CHANGES_REQUESTED** [review]

File: src/auth.py (hunk 1 above)
Comment (nogaems) [reply src/auth.py:42]: >You can pass the expiration time as a UTC UNIX timestamp (an int) or as a datetime, which will be converted into an int. 

This is what `jwt` documentation says.

File: src/auth.py (hunk 2 above)
Comment (nogaems) [reply src/auth.py:88]: You can read any issued `jwt` token contents anyways, it's just a signature system, not encrypting one.

File: src/util.py (hunk 3 above)
Comment (nogaems) [reply src/util.py]: Done.

**Review by gordon-quad — APPROVED** [review]

**Pull request merged**